lib/vrl/stdlib/src/chunks.rs

```diff
@@ -0,0 +1,153 @@
+use ::value::Value;
+use vrl::prelude::*;
+
+const MAX_UTF8_CODE_POINT_LENGTH_IN_BYTES: usize = 4;
+
+fn chunks(value: Value, chunk_size: Value) -> Resolved {
+    let chunk_size = chunk_size.try_integer()? as usize;
+    let bytes = value.try_bytes()?;
+
+    let data_length = bytes.len();
+    let mut chunked_data: Vec<&[u8]> = Vec::new();
+
+    let mut backtrack = 0;
+    let mut start = 0;
+    let mut end;
+
+    if chunk_size < MAX_UTF8_CODE_POINT_LENGTH_IN_BYTES {
+        return Err(r#""chunk_size" must be greater than or equal to 4 bytes"#.into());
+    }
+
+    loop {
+        let is_last_chunk = (start < data_length) && (start + chunk_size >= data_length);
+
+        if is_last_chunk {
+            chunked_data.push(&bytes[start..]);
+            break;
+        }
+
+        end = start + chunk_size - backtrack;
+
+        if is_root_utf8_byte(&bytes[end]) {
+            chunked_data.push(&bytes[start..end]);
+            backtrack = 0;
+            start = end;
+        } else {
+            backtrack += 1;
+
+            if backtrack >= MAX_UTF8_CODE_POINT_LENGTH_IN_BYTES {
+                return Err("Bytes are not valid UTF-8".into());
+            }
+        }
+    }
+
+    Ok(chunked_data.into_iter().collect::<Vec<_>>().into())
+}
+
+fn is_root_utf8_byte(byte: &u8) -> bool {
+    byte >> 6 != 0b10
+}
+
+#[derive(Clone, Copy, Debug)]
+pub struct Chunks;
+
+impl Function for Chunks {
+    fn identifier(&self) -> &'static str {
+        "chunks"
+    }
+
+    fn parameters(&self) -> &'static [Parameter] {
+        &[
+            Parameter {
+                keyword: "value",
+                kind: kind::BYTES,
+                required: true,
+            },
+            Parameter {
+                keyword: "chunk_size",
+                kind: kind::INTEGER,
+                required: true,
+            },
+        ]
+    }
+
+    fn examples(&self) -> &'static [Example] {
+        &[Example {
+            title: "chunks by byte",
+            source: r#"chunks("foobar", 1)"#,
+            result: Ok("hi"),
+        }]
+    }
+
+    fn compile(
+        &self,
+        _state: (&mut state::LocalEnv, &mut state::ExternalEnv),
+        _ctx: &mut FunctionCompileContext,
+        mut arguments: ArgumentList,
+    ) -> Compiled {
+        let value = arguments.required("value");
+        let chunk_size = arguments.required("chunk_size");
+
+        Ok(Box::new(ChunksFn { value, chunk_size }))
+    }
+}
+
+#[derive(Debug, Clone)]
+struct ChunksFn {
+    value: Box<dyn Expression>,
+    chunk_size: Box<dyn Expression>,
+}
+
+impl Expression for ChunksFn {
+    fn resolve(&self, ctx: &mut Context) -> Resolved {
+        let value = self.value.resolve(ctx)?;
+        let chunk_size = self.chunk_size.resolve(ctx)?;
+
+        chunks(value, chunk_size)
+    }
+
+    fn type_def(&self, _state: (&state::LocalEnv, &state::ExternalEnv)) -> TypeDef {
+        TypeDef::array(Collection::from_unknown(Kind::bytes())).infallible()
+    }
+}
```

Comment:
I'd love to know what `impl Expression` and `impl Function` do in the Vector codebase

Reply:
These are traits required by the compiler for each function. `Function` describes the function, and then `Expression` implements the evaluation logic.